hail/src/main/scala/is/hail/expr/ir/TableWriter.scala

```diff
@@ -433,7 +436,114 @@ case class TableTextWriter(
+    cb += os.invoke[Array[Byte], Unit]("write", const(headerStr).invoke[Array[Byte]]("getBytes"))
+    cb += os.invoke[Int, Unit]("write", '\n')
+  }
+
+  def consumeElement(cb: EmitCodeBuilder, element: EmitCode, os: Value[OutputStream], region: Value[Region]): Unit = {
+    require(element.st.virtualType == rowType)
+    val delimBytes: Value[Array[Byte]] = cb.memoize(cb.emb.getObject(delimiter.getBytes))
+
+    element.toI(cb).consume(cb, { cb._fatal("stream element can not be missing!") }, { case sv: SBaseStructValue =>
+      // I hope we're buffering our writes correctly!
+      (0 until sv.st.size).foreachBetween { i =>
+        val f = sv.loadField(cb, i)
+        val annotation = f.consumeCode[AnyRef](cb, Code._null[AnyRef],
+          { sv => StringFunctions.svalueToJavaValue(cb, region, sv) })
+        val str = Code.invokeScalaObject2[Any, Type, String](TableAnnotationImpex.getClass, "exportAnnotation",
+          annotation, cb.emb.getType(f.st.virtualType))
+        cb += os.invoke[Array[Byte], Unit]("write", str.invoke[Array[Byte]]("getBytes"))
+      }(cb += os.invoke[Array[Byte], Unit]("write", delimBytes))
+      cb += os.invoke[Int, Unit]("write", '\n')
+    })
+  }
+}
+
+case class TableTextFinalizer(outputPath: String, rowType: TStruct, delimiter: String,
+    header: Boolean = true, exportType: String = ExportType.CONCATENATED) extends MetadataWriter {
+  def annotationType: Type = TArray(TString)
+  def writeMetadata(writeAnnotations: => IEmitCode, cb: EmitCodeBuilder, region: Value[Region]): Unit = {
+    val ctx: ExecuteContext = cb.emb.ctx
+    val ext = ctx.fs.getCodecExtension(outputPath)
+    val partPaths = writeAnnotations.get(cb, "write annotations cannot be missing!")
+    exportType match {
+      case ExportType.CONCATENATED =>
+        val files = partPaths.castTo(cb, region, SJavaArrayString(true), false)
+        val jFiles = if (header) {
+          val headerFilePath = ctx.createTmpPath("header", ext)
+          val headerStr = rowType.fields.map(_.name).mkString(delimiter)
+          val os = cb.memoize(cb.emb.create(const(headerFilePath)))
+          cb += os.invoke[Array[Byte], Unit]("write", const(headerStr).invoke[Array[Byte]]("getBytes"))
+          cb += os.invoke[Int, Unit]("write", '\n')
+          cb += os.invoke[Unit]("close")
+
+          val jFiles = files.asInstanceOf[SJavaArrayStringValue].array
+          val allFiles = cb.memoize(Code.newArray[String](jFiles.length + 1))
+          cb += (allFiles(0) = const(headerFilePath))
+          cb += Code.invokeStatic5[System, Any, Int, Any, Int, Int, Unit](
+            "arraycopy", jFiles /*src*/, 0 /*srcPos*/, allFiles /*dest*/, 1 /*destPos*/, jFiles.length /*len*/)
+          allFiles
+        } else {
+          files.asInstanceOf[SJavaArrayStringValue].array
+        }
+
+        cb += cb.emb.getFS.invoke[Array[String], String, Unit]("concatenateFiles", jFiles, const(outputPath))
+        // can I rely on context managers to clean up?
```

Comment:
The only thing I'd say now is that we need to answer this question and delete this comment. I'm not sure what the answer is. What are you worried about cleaning up specifically?